### PR TITLE
fix(sdk): generate valid pipeline from non-ascii name. Fixes #5119

### LINKED ISSUE #5119: Invalid pipeline is generated if its name contains non-ascii chars only
### What steps did you take:
[A clear and concise description of what the bug is.]

Suppose there is a pipeline with non-ascii name:

```py
# sample.py
from kfp import dsl
@dsl.pipeline(name="サンプル")
def sample():
    ...
```

Then compile it:

```bash
dsl-compile --py sample.py --out sample.yaml
```

The sample.yaml becomes:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: Pipeline-
  ...
```

Then upload the pipeline and create run.

### What happened:

Failed because `Pipeline-` contains upper case char (`P`)

image

```json
{
   "error":"Failed to create a new run.: InternalServerError: Failed to create a workflow for (): Workflow.argoproj.io \"Pipeline-45z6j\" is invalid: [metadata.generateName: Invalid value: \"Pipeline-\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), metadata.name: Invalid value: \"Pipeline-45z6j\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]",
   "message":"Failed to create a new run.: InternalServerError: Failed to create a workflow for (): Workflow.argoproj.io \"Pipeline-45z6j\" is invalid: [metadata.generateName: Invalid value: \"Pipeline-\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), metadata.name: Invalid value: \"Pipeline-45z6j\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]",
   "code":13,
   "details":[
      {
         "@type":"type.googleapis.com/api.Error",
         "error_message":"Internal Server Error",
         "error_details":"Failed to create a new run.: InternalServerError: Failed to create a workflow for (): Workflow.argoproj.io \"Pipeline-45z6j\" is invalid: [metadata.generateName: Invalid value: \"Pipeline-\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), metadata.name: Invalid value: \"Pipeline-45z6j\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"
      }
   ]
}
```

### What did you expect to happen:

Successfully create run.

### Environment:

How did you deploy Kubeflow Pipelines (KFP)? AI Platform Pipelines

KFP version:  1.0.4

KFP SDK version:  1.3.0

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

The string `'Pipeline'` comes from 

https://github.com/kubeflow/pipelines/blob/3d40bba9a3e14b170a2c05e41a0b0bee64d399b8/sdk/python/kfp/compiler/compiler.py#L692

/kind bug

### PR DESCRIPTION
**Pull request by yuku**
**Description of your changes:**
Fixes #5119.

**Quality assurance**

Verify that the patched `dsl-compile` generates pipeline.yaml with `generateName: pipeline-`

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: pipeline-
  ...
```

and confirm I can create run for the newly uploaded pipeline.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -689,7 +689,8 @@ def _create_pipeline_workflow(self, parameter_defaults, pipeline, op_transformer
         exit_handler = first_group.exit_op
 
     # The whole pipeline workflow
-    pipeline_name = pipeline.name or 'Pipeline'
+    # It must valid as a subdomain
+    pipeline_name = pipeline.name or 'pipeline'
```
Comment (yuku) [inline sdk/python/kfp/compiler/compiler.py:693]: If `@dsl.pipeline(name='サンプル')` is given (nit: "サンプル" means sample in Japanese), the `pipeline_meta.name` becomes `"サンプル"` and `pipeline_name` becomes `""` in the ↓ line:

https://github.com/kubeflow/pipelines/blob/3d40bba9a3e14b170a2c05e41a0b0bee64d399b8/sdk/python/kfp/compiler/compiler.py#L832

Then a `Pipeline` instance is created with the empty string here:

https://github.com/kubeflow/pipelines/blob/3d40bba9a3e14b170a2c05e41a0b0bee64d399b8/sdk/python/kfp/compiler/compiler.py#L860

Then the `_create_pipeline_workflow()` is called with the `dsl_pipeline` object:

https://github.com/kubeflow/pipelines/blob/3d40bba9a3e14b170a2c05e41a0b0bee64d399b8/sdk/python/kfp/compiler/compiler.py#L890-L892

Finally, the given object becomes the `pipeline` variable in the changed line.

Comment (numerology) [inline sdk/python/kfp/compiler/compiler.py:693]: IMHO another way to fix this issue with less change to the current behavior is to first call `sanitize_k8s_name` to `pipeline.name` and if that's empty, we replace it with 'Pipeline'. 

WDYT?

**Review by numerology — COMMENTED** [review]
Thanks for the fix!

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (yuku) [reply sdk/python/kfp/compiler/compiler.py:693]: It is the current behavior. The string object assigned to `pipeline.name` has already been sanitized at [this line](https://github.com/kubeflow/pipelines/blob/3d40bba9a3e14b170a2c05e41a0b0bee64d399b8/sdk/python/kfp/compiler/compiler.py#L832). (https://github.com/kubeflow/pipelines/pull/5120#discussion_r573462700 describes the journey of the string. Please let me know if I missed anything.)

In addition, the `sanitize_k8s_name` returns an empty string if the given string only contains non-ascii char. So the default value is used only when the name of pipeline doesn't contain ascii char at all. I believe this change will never affect to at least English developers. 😉 

image

Comment (numerology) [reply sdk/python/kfp/compiler/compiler.py:693]: Ah got it.

**Comment by numerology** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**